plugin/edit.py

```diff
@@ -140,3 +150,190 @@ def _sort_by_application_order(changes: Iterable[TextEditTuple]) -> list[TextEdi
+    label: str = "Apply workspace edits", yes_button=""
+) -> Promise[sublime.DialogResult]:
+    changes = parse_workspace_edit(response, label)
+    file_count = len(changes.keys())
+    if file_count <= 1:
+        return session.apply_parsed_workspace_edits(changes, True).then(lambda _: sublime.DialogResult.YES)
+    total_changes = sum(map(lambda change: len(change[0]), changes.values()))
+    message = f"Apply {total_changes} changes across {file_count} files?"
+    choice = sublime.yes_no_cancel_dialog(message, yes_button, "Preview", title=label)
+    if choice == sublime.DialogResult.YES:
+        return session.apply_parsed_workspace_edits(changes, True).then(lambda _: sublime.DialogResult.YES)
+    elif choice == sublime.DialogResult.NO:
+        _render_workspace_edit_panel(session, response, changes, label, total_changes, file_count, on_accept_command)
+    return Promise.resolve(sublime.DialogResult.NO)
+
+
+BUTTONS_TEMPLATE = """
+<style>
+    html {{
+        background-color: transparent;
+        margin-top: 1.5rem;
+        margin-bottom: 0.5rem;
+    }}
+    a {{
+        line-height: 1.6rem;
+        padding-left: 0.6rem;
+        padding-right: 0.6rem;
+        border-width: 1px;
+        border-style: solid;
+        border-color: #fff4;
+        border-radius: 4px;
+        color: #cccccc;
+        background-color: #3f3f3f;
+        text-decoration: none;
+    }}
+    html.light a {{
+        border-color: #000a;
+        color: white;
+        background-color: #636363;
+    }}
+    a.primary, html.light a.primary {{
+        background-color: color(var(--accent) min-contrast(white 6.0));
+    }}
+</style>
+<body id='lsp-buttons'>
+    <a href='{apply}' class='primary'>Apply</a>&nbsp;
+    <a href='{discard}'>Discard</a>
+</body>"""
+
+DISCARD_COMMAND_URL = sublime.command_url('chain', {
+    'commands': [
+        ['hide_panel', {}],
+        ['lsp_hide_workspace_edit_buttons', {}]
+    ]
+})
+
+
+def _render_workspace_edit_panel(
+    session: Session,
+    workspace_edit: WorkspaceEdit,
+    changes_per_uri: WorkspaceChanges,
+    label: str,
+    total_changes: int,
+    file_count: int,
+    on_accept_command: tuple[str, dict] | None = None
+) -> None:
+    def _get_relative_path(wm: WindowManager, file_path: str) -> str:
+        base_dir = wm.get_project_path(file_path)
+        return os.path.relpath(file_path, base_dir) if base_dir else file_path
+
+    wm = windows.lookup(session.window)
+    if not wm:
+        return
+    pm = wm.panel_manager
+    if not pm:
+        return
+    panel = pm.ensure_workspace_edit_panel()
+    if not panel:
+        return
+    to_render: list[str] = []
+    reference_document: list[str] = []
+    header_lines = f"{total_changes} changes across {file_count} files - {label}\n"
+    to_render.append(header_lines)
+    reference_document.append(header_lines)
+    ROWCOL_PREFIX = " {:>4}:{:<4} {}"
+    base_dir = None
```

Comment:
the way how `base_dir` is set is a bit different in this PR.

Previously, `base_dir` was set by grabbing using the view, see:
https://github.com/sublimelsp/LSP/pull/2716/changes#diff-1624146a688a4081a8a84dab444a153add3a434c7617c30b019ab7039110bb9aL287

Now, base_dir is set when iterating through `changes_per_uri` as there we have a file path. (but I don't like the if in the for loop, I will see if I can do it a bit differently. If you have suggestions, please leave them)